core/os/thread.h

```diff
@@ -74,26 +79,25 @@ class Thread {
 private:
 	friend class Main;
 
-	static ID main_thread_id;
-
-	static uint64_t _thread_id_hash(const std::thread::id &p_t);
-
-	ID id = _thread_id_hash(std::thread::id());
+	ID id = UNASSIGNED_ID;
+	static SafeNumeric<uint64_t> id_counter;
 	static thread_local ID caller_id;
 	std::thread thread;
 
 	static void callback(ID p_caller_id, const Settings &p_settings, Thread::Callback p_callback, void *p_userdata);
 
 	static PlatformFunctions platform_functions;
 
+	static void make_main_thread() { caller_id = MAIN_ID; }
```

Comment:
IMO this could probably just be inlined into the callers

Reply:
I think this is ok as-is, code feels cleaner to me in intention and compilers will inline it anyway.